docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,622 @@
+> Custom cast types can be defined, we ship one: [`array`](https://github.com/discourse/discourse/blob/main/lib/active_support_type_extensions/array.rb).
+
+> 🙅 Don’t define attributes if you don’t transform them or validate them. The primary purpose of a contract is to validate data, it can also be used to cast or massage data before using it (usually a contract does both).
+
+Then, we define validations, exactly like you would in an `ActiveRecord` model. Here, we’re checking for `id` and `username` not being blank and that `username` respects an expected format.
+
+Another thing that is available in a contract, since it’s an `ActiveModel` object, are validation callbacks. If you need to manipulate the attribute values, you can do so by calling `before_validation` or `after_validation`. There are examples in the codebase, like in the [`Chat::CreateCategoryChannel`](https://github.com/discourse/discourse/blob/main/plugins/chat/app/services/chat/create_category_channel.rb#L42) service.
+
+Some methods have been added to the contract object to better mimic a classic hash. That way, when accessing/using `params` in a step, you don’t have to think too much about it:
+- `params.my_key` is also accessible through `params[:my_key]`.
+- `params.my_key = value` can also be done through `params[:my_key] = value`.
+- `#slice` and `#merge` are available.
+- `#to_hash` has been implemented, so the contract object will be automatically cast as a hash by Ruby depending on the context. For example, with an ActiveRecord model, you can do this: `user.update(**params)`.
+
+## Policy objects
+
+When a policy starts becoming complex or when you’d like to provide more context on why it can fail, then it’s time to use a policy object instead of a simple policy.
+
+It’s quite easy to create a new policy object, let’s take the `can_update_username` policy we have in the `User::UpdateUsername` service and convert it:
+```ruby
+class User::Policy::CanUpdateUsername < Service::PolicyBase
+  delegate :user, to: :context, private: true
+
+  def call
+    guardian.can_edit_username?(user)
+  end
+
+  def reason
+    # Here we can put more complex logic to dynamically output a reason, this is just an example
```

Comment:
maybe we should show a slightly more complex at least accessing user to show how you can use the variables available